docs/compactor.md

```diff
@@ -0,0 +1,93 @@
+
+* Full-Snapshot-revision-1
+* Full-Snapshot-revision-18
+
+Then, during restoration Full-Snapshot-revision-18, and all the delta snaphots from Incremental-Snapshot-revision-22-18 to Incremental-Snapshot-revision-52-48 will be applied. **Restorer would restore from latest ready full snapshot and all subsequent delta snapshots.** As Full-Snapshot-revision-38 is not present in the metadata file, it would not be considered as ready yet.
+
+Another approach to make sure that restorer only picks up snapshots which are ready, full snapshots can be named with extension `.part` until the snapshot is ready. When the snapshot is ready, `.part` extension can be removed from it's name. Restorer would avoid any file having extension as `.part`.  
+
+So, if Full-Snapshot-revision-38 is not ready in the above example, it can be named as `Full-Snapshot-revision-38.part`. Restorer will apply Full-Snapshot-revision-18 and all the delta snaphots from Incremental-Snapshot-revision-22-18 to Incremental-Snapshot-revision-52-48. **Restorer would restore from latest ready full snapshot and all subsequent delta snapshots.**
+
+This option is preferred as it avoids introducing a metadata file. But it would work only if renaming to remove the `.part` extension (upon completion of upload) is an atomic operation in all the supported snapstores. This needs to be evaluated.
+
+#### Backward Compatibility
+1. **Restoration** : The changes to handle the newly proposed backup directory structure must be backward compatible with older structures at least for restoration because we need have to restore from backups in the older structure. This includes the support for restoring from a backup without a metadata file if that is used in the actual implementation.
+2. **Backup** : For new snapshots (even on a backup containing the older structure), the new structure may be used. The new structure must be setup automatically including creating the base full snapshot (and also the creation of a metadata file if necessary).
+3. **Garbage collection** : The existing functionality of garbage collection of snapshots (full and incremental) that fall out of the backup retention policy must be compatible with both old and new backup folder structure. I.e. the snapshots in the older backup structure must be retained in their own structure and the snapshots in the proposed backup structure should be retained in the proposed structure. Once all the snapshots in the older backup structure go out of the retention policy and are garbage collected, we can think of removing the support for older backup folder structure.  
+
+**Note:** ETCD Backup Restore needs to take one full snapshot just after starting. Compactor will work only if there is any full snapshot already present in the store. It is not limitation but a design choice
```

Comment:
Hmm... I know we do that, but hasn't that already caused trouble?

/cc @amshuman-kr @shreyas-s-rao 

I don't really see why this is necessary/beneficial. In the contrary, this seems problematic. Right when we need ETCD the most, i.e. when it starts and all the API server start hammering it, we also do a full snapshot. What was the motivation?

Wouldn't it make more sense to only do that if:
- There is no full snapshot at all (we need at least one for the delta snapshots to make sense, but not a new one every time ETCD starts)
- The last one is "too old"

Reply:
On a call out-of-band, it was said we don't take full snapshot always, but only if we miss one or it's older than 24h.

Let's then add this info to the "note" above.

Reply:
@abdasgupta Can you please add the information to the doc?